### ISSUE TITLE
Disallow executable bit on .nix files

### ISSUE THREAD

**Issue by patka-123** [NixOS/nixpkgs-vet#110]
I came across [a nixpkgs PR](https://github.com/NixOS/nixpkgs/pull/341771) that removes the executable bit from some .nix files. This can be prevented by a check in CI.

Would this project (nixpkgs-vet) be appropriate for adding such a check, or would it fit better in a standalone loose action?

If we decide where it should go I'd like to take a stab at it.

**Comment by infinisil** [comment]
Yeah this would be a pretty good fit for such a check! In particular it would fit around here: https://github.com/NixOS/nixpkgs-vet/blob/0d26da8aa2bbc5d74461ca72391d6260ea51d028/src/structure.rs#L170-L176

I think for this case it might be fine to just add a simple additional check and error, but in theory this _could_ break master when a PR introduces an executable file but already went through CI before this change. So whenever strictness of checks is increased, a [ratchet check](https://github.com/NixOS/nixpkgs-vet/tree/main?tab=readme-ov-file#ratchet-checks) would be most ideal to prevent such situations.

If you do want to go for the full ratchet check, you'd have to extend [`ratchet::Package`](https://github.com/NixOS/nixpkgs-vet/blob/0d26da8aa2bbc5d74461ca72391d6260ea51d028/src/ratchet.rs#L34-L40) with a something like a `RatchetState<Executable>`, change [`check_package`](https://github.com/NixOS/nixpkgs-vet/blob/0d26da8aa2bbc5d74461ca72391d6260ea51d028/src/structure.rs#L119-L125) to return `validation::Result<HashMap<String, RatchetState<Executable>>>`, and then incorporate those `RatchetState<Executable>` values into the construction of `ratchet::Package` in [`eval.rs`](https://github.com/NixOS/nixpkgs-vet/blob/0d26da8aa2bbc5d74461ca72391d6260ea51d028/src/eval.rs#L154) (or a refactoring to move the construction out of that file).

While this would be neat, no offense taken if you don't go for that, a non-ratchet check is fine too (and perhaps somebody from @NixOS/nixpkgs-vet would be interested in making it a ratchet one) :)

**Comment by infinisil** [comment]
> Does it really break master though? A nix file with an executable bit falling through the cracks doesn't seem like the end of the world to me.

Scenario:
- There's a PR that adds an executable file, CI is successful because there's no check preventing that just yet
- We implement a check to ensure there are no executables, but it only runs for PRs that get updated
- The PR gets merged without updating it. CI is still green, so nothing seems wrong with doing so, but it breaks master, because new any updated PR will fail on the new executable file

The premise of the ratchet checks is to avoid such a situation, by checking the base branch _against_ the PR branch, only ensuring that _new_ instances of a bad pattern aren't introduced, not that there is no such pattern at all.

Granted, it's hard to say exactly what the chance is of this occurring. The fact that there were like 10 executable files before makes be believe it's not too uncommon, but maybe not worth creating a ratchet check for. Ratchet checks become less useful the less instances of a problematic pattern there are, but it's hard to draw the line.

At some point I would like to have the code in a state where it's just as easy to add a new ratchet check as it is to add a strict check. We're not there just yet, but what pattern works should become more obvious as more ratchet checks are implemented.

Anyways, I'd be happy to get a PR even for the basic check as a start :)

**Comment by willbush** [comment]
FWIW checking executable bit on current master (still?) returns nothing:

```sh
$ find . -type f -name "*.nix" -executable -print
```

**Comment by infinisil** [comment]
@willbush They were removed somewhat recently, I guess nobody added any since: https://github.com/NixOS/nixpkgs/pull/341771

**Comment by willbush** [comment]
@infinisil still think this should be a ratchet check?

**Comment by infinisil** [comment]
Not worth it for now imo, but hopefully at some point ratchet checks are as easy as non-ratchet ones :)

**Comment by philiptaron** [comment]
##144 moves us in that direction!

**Comment by philiptaron** [comment]
> There are valid reasons for .nix files to be executable, when they have a she-bang. A more fine-grained approach would be to force `+x` on all files with she-bang and disallow it on all others.

Agree: that’s the implementation we want.
